flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java

```diff
@@ -0,0 +1,237 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import static java.util.Collections.singletonList;
+import static org.apache.flink.connector.pulsar.common.utils.PulsarClosingUtils.closeWithTimeout;
+
+/** The enumerator class for pulsar source. */
+@Internal
+public class PulsarSourceEnumerator
+        implements SplitEnumerator<PulsarPartitionSplit, PulsarSourceEnumState> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PulsarSourceEnumerator.class);
+
+    /** The admin interface for pulsar. */
+    private final PulsarAdmin pulsarAdmin;
+
+    private final PulsarSubscriber subscriber;
+    private final RangeGenerator rangeGenerator;
+    private final Configuration configuration;
+    private final SourceConfiguration sourceConfiguration;
+    private final SplitEnumeratorContext<PulsarPartitionSplit> context;
+    private final SplitsAssignmentState assignmentState;
+
+    public PulsarSourceEnumerator(
+            PulsarAdmin pulsarAdmin,
+            PulsarSubscriber subscriber,
+            RangeGenerator rangeGenerator,
+            SerializableSupplier<StartCursor> startCursorSupplier,
+            SerializableSupplier<StopCursor> stopCursorSupplier,
+            Configuration configuration,
+            SourceConfiguration sourceConfiguration,
+            SplitEnumeratorContext<PulsarPartitionSplit> context) {
+        this.pulsarAdmin = pulsarAdmin;
+        this.subscriber = subscriber;
+        this.rangeGenerator = rangeGenerator;
+        this.configuration = configuration;
+        this.sourceConfiguration = sourceConfiguration;
+        this.context = context;
+        this.assignmentState =
+                new SplitsAssignmentState(
+                        startCursorSupplier, stopCursorSupplier, sourceConfiguration);
+    }
+
+    public PulsarSourceEnumerator(
+            PulsarAdmin pulsarAdmin,
+            PulsarSubscriber subscriber,
+            RangeGenerator rangeGenerator,
+            SerializableSupplier<StartCursor> startCursorSupplier,
+            SerializableSupplier<StopCursor> stopCursorSupplier,
+            Configuration configuration,
+            SourceConfiguration sourceConfiguration,
+            SplitEnumeratorContext<PulsarPartitionSplit> context,
+            PulsarSourceEnumState sourceEnumState) {
+        this.pulsarAdmin = pulsarAdmin;
+        this.subscriber = subscriber;
+        this.rangeGenerator = rangeGenerator;
+        this.configuration = configuration;
+        this.sourceConfiguration = sourceConfiguration;
+        this.context = context;
+        this.assignmentState =
+                new SplitsAssignmentState(
+                        startCursorSupplier,
+                        stopCursorSupplier,
+                        sourceConfiguration,
+                        sourceEnumState);
+    }
+
```

Comment:
Please merge ctors for example by adding `SplitsAssignmentState` as a parameter and filling that in at call-site.